### PR TITLE
introducing get_tag and math sigmoid

### PR DESCRIPTION
**Pull request by ohhmm**

### CODE REVIEW

**Comment by eldiener** [comment]
Please explain the purpose and functionality of these additions. Also why you think they should be part of the MPL library.

**Comment by ohhmm** [comment]
hi @eldiener, get_tag is for lazy getting of member definition of an incomplete definition and
sigmoid calculates rational sigmoid that may be useful for compile-time sigmoid calculation for compile-time neural network definitions.

**Comment by eldiener** [comment]
The get_tag might be useful in MPL but sigmoid belongs in some other place. MPL should not depend on the other Boost libraries needed by sigmoid and MPL is not a math library.

**Comment by ohhmm** [comment]
Updated.

**Comment by eldiener** [comment]
The code uses C++11. It will fail if used in C++03. Is that what you want ? I also need the appropriate documentation for it added to the MPL documentation and at least some simple test, showing its works as expected, added to the MPL tests.

**Comment by ohhmm** [comment]
Updated.

**Comment by eldiener** [comment]
I think the struct generated by the BOOST_MPL_GET_TAG_DEF macro should be in the boost::mpl namespace rather than in a global namespace. Also would not your test be better if it made sure the type generated was the same as the nested type of the T ? I realize the latter assertion seems trivial but it would validate the construct generated by the BOOST_MPL_GET_TAG_DEF macro, which is really what your test wants to do.

**Comment by ohhmm** [comment]
Updated.

File: doc/refmanual/get-tag-def.html [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+#define <a href="./has-xxx-trait-def.html" class="identifier">BOOST_MPL_GET_TAG_DEF</a>(name) \
+    <em>unspecified token sequence</em> \
+/**/
+</pre>
+</div>
+<div class="section" id="id1330">
+<h3><a class="subsection-title" href="#description" name="description">Description</a></h3>
+<p>Expands into a definition of a lazy getter <a class="reference internal" href="./metafunction.html">Metafunction</a> <tt class="literal"><span class="pre">get_tag_name</span></tt>
+such that for any type <tt class="literal"><span class="pre">x</span></tt> <tt class="literal"><span class="pre">get_tag_name&lt;x&gt;::type</span> <span class="pre">is</span> <span class="pre">type</span></tt> where
+ <tt class="literal"><span class="pre">name</span></tt> is a nested type memeber of <tt class="literal"><span class="pre">x</span></tt>.</p>
```
Comment (eldiener) [inline doc/refmanual/get-tag-def.html]: Spelling error: "memeber" should be "member".

**Comment by eldiener** [comment]
Still needed is an entry in the list of Introspection macros, else no one will find the macro in the MPL docs. I also added a comment about a misspelling in the html page for the macro. Other than these two doc fixes everything else looks good.

I am still not sure whether your implementation really belongs as part of a template metaprogamming library but I will add it as it might be useful to others, and I see no harm in including it in MPL as a further technical aid.

**Comment by ohhmm** [comment]
Updated.

**Comment by eldiener** [comment]
The documentation for BOOST_MPL_GET_TAG_DEF needs to be updated to reflect the fact that its expansion is in the boost::mpl namespace.

**Comment by ohhmm** [comment]
Updated.

**Pull request merged**